solutions/gold/usaco-597.mdx

```diff
@@ -17,6 +35,132 @@ The second and third solutions in the analysis fail on the following test case:
 1
 ```
 
-The second one gives the wrong answer, and the third gives runtime error.
-
+The second one gives the wrong answer, and the third errors.
 </Warning>
+
+From the hints, we'll start binary searching on the initial power $R$. Now, we've reduced the problem into checking if a fixed power $R$ works for our set of haybales.
+
+There are many ways to approach this (see the [Official Analysis](http://www.usaco.org/current/data/sol_angry_gold_jan16.html) for more information!), but perhaps the most straightforward way is running a second binary search on the maximum point we can place our cow.
+
+This works because just like our power, there exists a set point where a cow being launched at a coordinate will not knock out all the haybales to the left/right. Since $N \leq 5 \cdot 10^4$, we can simulate these explosions naively.
+
+## Implementation
+
+**Time Complexity**: $\mathcal{O}(N \log^2 N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+#define all(x) begin(x), end(x)
+
+const int MAX_POS = 1e9;
+
+int main() {
+	freopen("angry.in", "r", stdin);
+	freopen("angry.out", "w", stdout);
+
+	int n;
+	cin >> n;
+
+	vector<int> bales(n);
+
+	for (int i = 0; i < n; i++) {
+		cin >> bales[i];
+		// to avoid floats, double all coordinates
+		bales[i] *= 2;
+	}
+
+	sort(all(bales));
+
+	/**
+	 * @returns whether it's possible to explode
+	 * the rest of the haybales before/after idx
+	 *
+	 * @param pos position of our cow
+	 * @param idx furthest haybale we hit with our radius
+	 * @param r current radius at our position
+	 * @param dir whether we're checking positions to the right/left
+	 */
+	function<bool(int, int, int, bool)> push = [&](int pos, int idx, int r,
```

Comment:
why not just make this global lmao? it's not like there are multiple test cases

Reply:
avoid global variables
and tbh it doesn't take away too much from readability